### PR TITLE
Arc 1200 web set up ga4 ua to bigquery pipeline

### PR DESCRIPTION
**Pull request by Frydafly**
# Pull Request for dbt-arc-functions

- [x] Create branch of dbt-arc-functions using JIRA extension so that naming convention is consistent

- [x] If you've created a macro, make sure it is documented by running the notebook: `create_docs.ipynb` util in this repo. From there, you'll have to either fill in descriptions by hand OR use our `fill_in_descriptions_using_openai.ipynb` to fill in descriptions using OpenAI's ChatGPT. Always make sure to check over ChatGPT's column descriptions, as they're likely not perfect.

- [x] Test branch on a development branch of an existing client (ideally the one that raised the bug). Do this by changing revision of the package to the branch name in `packages.yml` file in dbt State client below.
bsd-nyccfb

- [x] If new macro files were created or updated, re-run `create_or_update_standard_models.py` for the client dbt project, replacing models and dependencies

- [x] check in dbt that client package successfully runs `dbt deps`, `dbt compile`, and `dbt run`; screenshot this page
Screenshot 2023-06-15 at 3 03 41 PM

- [x] query the resulting staging model in bigquery (or by previewing mart in dbt cloud, or your IDE); screenshot this
this one is for the mart that reports by day and by session
Screenshot 2023-06-09 at 5 17 35 PM

this one is the daily rollup that counts unique sessions per day
Screenshot 2023-06-09 at 5 18 21 PM

## Formatters

Three different formatters will run when you open a pull request:
- [sqlfmt](http://sqlfmt.com/)
- [yamlfix](https://lyz-code.github.io/yamlfix/)
- [Black(python)](https://pypi.org/project/black/)

If any of them detect syntax problems, they will error. They will automatically open a pull request with the changes you need to make to pass the formatter. Just merge that PR and you should be good to go.

## Linters
We have two in-house linters:

### [check_for_docs_for_every_macro.py](https://github.com/bsd/dbt-arc-functions/blob/main/utils/check_for_docs_for_every_macro.py)
Just makes sure that every macro has a doc associated with it in the documentation folder with the same folder structure, i.e.:

`documentation/combined_email_paidmedia/marts/mart_combined_email_paidmedia_daily_revenue_performance.yml`

is the doc for

`       macros/combined_email_paidmedia/marts/mart_combined_email_paidmedia_daily_revenue_performance.sql`

### [check_docs_correct.py](https://github.com/bsd/dbt-arc-functions/blob/main/utils/check_docs_correct.py)
Runs through a lot of checks on documents to make sure that they are formatted to our standards. This includes checking whether docs and sources:
- have columns
- have tables
- source columns have data_type and description
- have a version number
- docs have a macro associated with them
- docs are not blank

There may be additional checks not documented here. The linter will fail on any error.

The script will document all failures and display them to you with remedies.

### CODE REVIEW

**Comment by Frydafly** [comment]
ready for review, but not ready for merge. pending passing the documentation linter, and also pending data validation on @jiffar1 end

File: sources/ga4_google_analytics_web.yml [hunk 1]
```diff
@@ -1,71 +1,305 @@
...
-# TODO: figure out how to pull in the analytics schema name
 version: 2
 sources:
   - name: ga4_google_analytics_web
     database: "{{ var('database') }}"
+    schema: '{{ target.name }}_ga4'
+    description: Google Analytics data for website
     tables:
-      - name: events
-        identifier: events_*
```
Comment (ryantimjohn) [inline sources/ga4_google_analytics_web.yml:10]: @Frydafly is this identifier no longer needed?

File: utils/Pipfile [hunk 2]
```diff
@@ -7,6 +7,10 @@ name = "pypi"
 "ruamel.yaml" = "==0.16.12"
 google-cloud-bigquery = "==2.31.0"
 click = "*"
+tiktoken = "*"
+pyyaml = "*"
```
Comment (ryantimjohn) [inline utils/Pipfile:11]: pyyaml used to conflict with ruamel.yaml but luckily they now have different namespaces!
https://pyyaml.org/wiki/PyYAMLDocumentation
https://yaml.readthedocs.io/en/latest/basicuse.html

**Review by ryantimjohn — APPROVED** [review]
Overall, everything looks good in Fryda's testing and through the code; will let people who know GA4 better comment on those particulars.

File: sources/ga4_google_analytics_web.yml (hunk 1 above)
Comment (Frydafly) [reply sources/ga4_google_analytics_web.yml:10]: we are no longer pulling in the events table -- it's done through configuration of the velir/ga4 package instead, so we don't need it to be part of the source anymore.

**Comment by Frydafly** [comment]
my only change after review was to fix merge conflicts, and then run yamlfix since all the merge conflicts were from the source file documentation that were in conflict with the main branch.

**Comment by Frydafly** [comment]
Thanks for the review. I'll hold on merging until I get confirmation from @jiffar1 that he is seeing what he expects when these changes were applied to bsd-nyccfb.

**Comment by Frydafly** [comment]
@jiffar1 wrote up regex to create a more appropriate field for grouping by page path which stops at the first directory 

``` sql
CASE
     WHEN page_path = "/" THEN "/"
     ELSE REGEXP_EXTRACT(page_path, "^(/[^/]+)")
 END as first_directory
```

I will be incorporating this field to the reporting mart.

**Comment by Frydafly** [comment]
Adding two pieces of feedback from @jiffar1: 


1. the default channel grouping shows (none) and (other) in BQ version, but does not show similarly in the direct GA4. Do you know where (none) and (other) are coming from? I’m hoping it would match the GA4 format if possible.

3. In the direct GA4 dashboard there are ‘Sessions' and 'Sessions per user’ columns. Do you know how we can recreate sessions per user in the BQ table? At present I’m only seeing sessions equaling unique users. So a way to count multiple sessions from the same user would allow that. (ps: I currently have it showing page views per unique user).

**Review by ryantimjohn — APPROVED** [review]
Looks good!

**Pull request merged**